include/groonga/string.h

```diff
@@ -58,7 +58,10 @@ GRN_API grn_rc grn_str_close(grn_ctx *ctx, grn_str *nstr);
 
 #define GRN_CHAR_BLANK 0x80
 #define GRN_CHAR_IS_BLANK(c) ((c) & (GRN_CHAR_BLANK))
-#define GRN_CHAR_TYPE(c) ((c) & 0x7f)
+#define GRN_REMOVED_CHAR_TYPE_SYMBOL 0x40
```

Comment:
Why do we need this new tag?

Reply:
> Why do we need this new tag?

To decide whether we output `|blank` or `|symbol` at https://github.com/groonga/groonga/pull/1312/files#diff-caf1d936b395dcac087bd2b6d8585de0e06695cfe00c899d9299dc9cfec2a118R33-R40, we need to add `GRN_REMOVED_CHAR_TYPE_SYMBOL`.

Reply:
Why do we need to output `|symbol` there?

Reply:
`WITH_TYPES` flag output a type of character after normalization.
At that time, I thought I should also output a type of character that is removed by normalization same as `remove_blank`.

In this case, the symbols like '#', '!', and so on are removed by `remove_symbol` option. Therefore, I implemented that Groonga output `|symbol` with a removed symbol by `remove_symbol` option.

Reply:
`GRN_CHAR_BLANK` is used in `grn_pat_scan()` (that is used for highlighting keywords) to detect suitable position.
But this new tag isn't used anywhere. (It's just used for inspection.)
If we should treat removed symbols like removed blanks, we can just use `GRN_CHAR_BLANK` for removed symbols too. (Or we may need to rename `GRN_CHAR_BLANK` to more suitable name.)
If we need special handling only for removed symbols like `GRN_CHAR_IS_BLANK()` related code in `grn_pat_can()`, we also need to implement it.

Reply:
I got it.
We should treat removed symbols like removed blanks, otherwise the positional shift of a highlight.

Therefore, we use GRN_CHAR_BLANK(or a more suitable name) for removed symbols too.

Reply:
I'm not sure whether we need `GRN_CHAR_BLANK` related behavior for `grn_pat_scan()`.
Let's try `highlight_html()` with this options before we change the current implementation.

Reply:
I tried `highlight_html()` with this `remove_symbol`.
The result was not correct as below.

```
table_create Entries TABLE_NO_KEY
[[0,1642570456.140369,0.003215312957763672],true]
column_create Entries body COLUMN_SCALAR ShortText
[[0,1642570456.144203,0.008811712265014648],true]

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenBigram --normalizer 'NormalizerNFKC130("remove_symbol", true)'
[[0,1642570456.154353,0.002437114715576172],true]
column_create Terms document_index COLUMN_INDEX|WITH_POSITION Entries body
[[0,1642570456.156936,0.004312515258789062],true]

load --table Entries
[
{"body": "Mroonga is a Ｍｙ ＳＱＬ storage engine based on Groonga. <b>Rroonga</b> is a Ruby binding of Groonga."}
]
[[0,1642570456.161346,0.001099348068237305],1]

select Entries \
  --match_columns body \
  --query 'groonga' \
  --output_columns 'highlight_html(body)' \
  --command_version 2
[[0,1642570568.166593,0.0003690719604492188],[[[0],[["highlight_html",null]]]]]
```

I tried using `GRN_CHAR_BLANK` for removed symbols too as below and execute `highlight_html()` with the following implementation.

```diff
diff --git a/lib/normalizer.c b/lib/normalizer.c
index 9cff26099..7b5cf212e 100644
--- a/lib/normalizer.c
+++ b/lib/normalizer.c
@@ -1796,11 +1796,11 @@ grn_nfkc_normalize(grn_ctx *ctx,
                                                   current,
                                                   current_length)) {
           if (context->t > context->types) {
-            if (data.options->char_type_func(current) == GRN_CHAR_SYMBOL) {
-              context->t[-1] |= GRN_REMOVED_CHAR_TYPE_SYMBOL;
-            } else {
+            //            if (data.options->char_type_func(current) == GRN_CHAR_SYMBOL) {
+            //              context->t[-1] |= GRN_REMOVED_CHAR_TYPE_SYMBOL;
+            //            } else {
               context->t[-1] |= GRN_CHAR_BLANK;
-            }
+              //            }
           }
           if (!data.options->include_removed_source_location) {
             source_ += current_length;
```

The result was correct as below.

```
table_create Entries TABLE_NO_KEY
[[0,1642571276.485663,0.003362178802490234],true]
column_create Entries body COLUMN_SCALAR ShortText
[[0,1642571276.489689,0.02276062965393066],true]

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenBigram --normalizer 'NormalizerNFKC130("remove_symbol", true)'
[[0,1642571276.512718,0.001276493072509766],true]
column_create Terms document_index COLUMN_INDEX|WITH_POSITION Entries body
[[0,1642571276.514189,0.01355886459350586],true]

load --table Entries
[
{"body": "Mroonga is a Ｍｙ ＳＱＬ storage engine based on Groonga. <b>Rroonga</b> is a Ruby binding of Groonga."}
]
[[0,1642571276.527843,0.001530885696411133],1]

select Entries \
  --match_columns body \
  --query 'groonga' \
  --output_columns 'highlight_html(body)' \
  --command_version 2
[[0,1642571276.529789,0.02620649337768555],[[[1],[["highlight_html",null]],["Mroonga is a Ｍｙ ＳＱＬ storage engine based on <span class=\"keyword\">Groonga</span>. &lt;b&gt;Rroonga&lt;/b&gt; is a Ruby binding of <span class=\"keyword\">Groonga</span>."]]]]
```

Therefore, we need to use GRN_CHAR_BLANK(or a more suitable name) for removed symbols too.

Reply:
The former example doesn't hit any records.
It seems that it's a search problem not a `highlight_html` problem.

Reply:
I mistake the method of confirmation.
I should confirm whether `remove_symbol` needs `GRN_CHAR_BLANK` or not.
Therefore, I need to confirm whether `highlight_html()` works correctly or not when I don't set `GRN_CHAR_BLANK`.

The above examples are not correct in this point of view.
Because the first example uses `GRN_REMOVED_CHAR_TYPE_SYMBOL` and the second example uses `GRN_CHAR_BLANK`.

I tried `highlight_html()` with `remove_symbol` without `GRN_REMOVED_CHAR_TYPE_SYMBOL` and `GRN_CHAR_BLANK` as below.

```diff
diff --git a/lib/normalizer.c b/lib/normalizer.c
index 9cff26099..22cd8a0af 100644
--- a/lib/normalizer.c
+++ b/lib/normalizer.c
@@ -1797,7 +1797,8 @@ grn_nfkc_normalize(grn_ctx *ctx,
                                                   current_length)) {
           if (context->t > context->types) {
             if (data.options->char_type_func(current) == GRN_CHAR_SYMBOL) {
-              context->t[-1] |= GRN_REMOVED_CHAR_TYPE_SYMBOL;
+              ;
+              //              context->t[-1] |= GRN_REMOVED_CHAR_TYPE_SYMBOL;
             } else {
               context->t[-1] |= GRN_CHAR_BLANK;
             }
```

The result was correct as below.

```
table_create Entries TABLE_NO_KEY
column_create Entries body COLUMN_SCALAR ShortText

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenBigram --normalizer 'NormalizerNFKC130("remove_symbol", true)'
column_create Terms document_index COLUMN_INDEX|WITH_POSITION Entries body

load --table Entries
[
{"body": "Groonga. Mroonga. Pgroonga."}
]

select Entries \
  --match_columns body \
  --query 'pgroonga' \
  --output_columns 'highlight_html(body)' \
  --command_version 2
[[0,1642580839.389963,0.06885743141174316],[[[1],[["highlight_html",null]],["Groonga. Mroonga. <span class=\"keyword\">Pgroonga</span>."]]]]
```

Therefore, we don't need to treat `removed_symbol` like `removed_blank`.
`GRN_REMOVED_CHAR_TYPE_SYMBOL` and `GRN_CHAR_BLANK` are needless in `remove_blank`.